api/service/stagedsync/sync_config.go

```diff
@@ -0,0 +1,378 @@
+	pb "github.com/harmony-one/harmony/api/service/legacysync/downloader/proto"
+	"github.com/harmony-one/harmony/core/types"
+	"github.com/harmony-one/harmony/internal/utils"
+)
+
+// Constants for syncing.
+const (
+	downloadBlocksRetryLimit        = 10 // downloadBlocks service retry limit
+	RegistrationNumber              = 3
+	SyncingPortDifference           = 3000
+	inSyncThreshold                 = 0   // when peerBlockHeight - myBlockHeight <= inSyncThreshold, it's ready to join consensus
+	SyncLoopBatchSize        uint32 = 30  // maximum size for one query of block hashes
+	verifyHeaderBatchSize    uint64 = 100 // block chain header verification batch size (not used for now)
+	LastMileBlocksSize              = 50
+
+	// after cutting off a number of connected peers, the result number of peers
+	// shall be between numPeersLowBound and numPeersHighBound
+	NumPeersLowBound  = 3
+	numPeersHighBound = 5
+	NumPeersReserved  = 2
+
+	downloadTaskBatch = 5
+)
+
+// SyncPeerConfig is peer config to sync.
+type SyncPeerConfig struct {
+	ip          string
+	port        string
+	peerHash    []byte
+	client      *downloader.Client
+	blockHashes [][]byte       // block hashes before node doing sync
+	newBlocks   []*types.Block // blocks after node doing sync
+	mux         sync.Mutex
+	failedTimes uint64
+}
+
+// CreateTestSyncPeerConfig used for testing.
+func CreateTestSyncPeerConfig(client *downloader.Client, blockHashes [][]byte) *SyncPeerConfig {
+	return &SyncPeerConfig{
+		client:      client,
+		blockHashes: blockHashes,
+	}
+}
+
+// GetClient returns client pointer of downloader.Client
+func (peerConfig *SyncPeerConfig) GetClient() *downloader.Client {
+	return peerConfig.client
+}
+
+// IsEqual checks the equality between two sync peers
+func (peerConfig *SyncPeerConfig) IsEqual(pc2 *SyncPeerConfig) bool {
+	return peerConfig.ip == pc2.ip && peerConfig.port == pc2.port
+}
+
+// GetBlocks gets blocks by calling grpc request to the corresponding peer.
+func (peerConfig *SyncPeerConfig) GetBlocks(hashes [][]byte) ([][]byte, error) {
+	response := peerConfig.client.GetBlocksAndSigs(hashes)
+	if response == nil {
+		return nil, ErrGetBlock
+	}
+	return response.Payload, nil
+}
+
+func (peerConfig *SyncPeerConfig) registerToBroadcast(peerHash []byte, ip, port string) error {
+	response := peerConfig.client.Register(peerHash, ip, port)
+	if response == nil || response.Type == pb.DownloaderResponse_FAIL {
+		return ErrRegistrationFail
+	} else if response.Type == pb.DownloaderResponse_SUCCESS {
+		return nil
+	}
+	return ErrRegistrationFail
+}
+
+// CompareSyncPeerConfigByblockHashes compares two SyncPeerConfig by blockHashes.
+func CompareSyncPeerConfigByblockHashes(a *SyncPeerConfig, b *SyncPeerConfig) int {
+	if len(a.blockHashes) != len(b.blockHashes) {
+		if len(a.blockHashes) < len(b.blockHashes) {
+			return -1
+		}
+		return 1
+	}
+	for id := range a.blockHashes {
+		if !reflect.DeepEqual(a.blockHashes[id], b.blockHashes[id]) {
```

Comment:
You do same check twice, better only
`bytes.Compare(a.blockHashes[id], b.blockHashes[id])`

Reply:
The first check is just about length and it helps performance.

Reply:
Why it should be different if it is hashes?

Reply:
It's array of byte arrays. So rather than going through them one by one, if length is not same (which is the reason for most cases), that means they are not equal. second loop goes through the array of byte arrays and check them one by one. for each of the items, we use bytes.Compare.